docs/topics/object-declarations.md

```diff
@@ -30,42 +60,59 @@ val ab: A = object : A(1), B {
 }
 ```
 
-If you need just an object, with no nontrivial supertypes, write:
+### Using anonymous object as return and value types
+
+When an anonymous object is used as a type of a local or [private](visibility-modifiers.md#packages) but not [inline](inline-functions.md)
+declaration (function or property), all its members are accessible via this function or property:
 
 ```kotlin
-fun foo() {
-    val adHoc = object {
-        var x: Int = 0
-        var y: Int = 0
+class C {
+    private fun getObject() = object {
+        val x: String = "x"
+    }
+
+    fun printX() {
+        println(getObject().x)
     }
-    print(adHoc.x + adHoc.y)
 }
 ```
 
-Note that anonymous objects can be used as types only in local and [private](visibility-modifiers.md#packages) declarations. If you use an anonymous object as a
-return type of a public function or the type of a public property, the actual type of that function or property
-will be the declared supertype of the anonymous object, or `Any` if you haven't declared any supertype. Members added
-in the anonymous object will not be accessible.
+If this function or property is public or private inline, its actual type is:
+* `Any` if the anonymous object doesn't have a declared supertype
+* the declared supertype of the anonymous object if there is exactly one such type 
+* the explicitly declared type if there is more than one declared supertype
+
+In all these cases, members added in the anonymous object are not accessible. Overriden members are accessible if they 
+are declared in the actual type of the function or property.
```

Comment:
As far as I could understand, the following example extends this sentence, so maybe we should put the colon.